### PR TITLE
Ofir/move live session

### PR DESCRIPTION
**Pull request by OfirZinkerz**

### CODE REVIEW

File: demo/calls/demoApp.js [hunk 1]
```diff
@@ -1,21 +1,24 @@
 (function (angular) {
     'use strict';
 
+    var isTeacher = localStorage.getItem('isTeacher');
```
Comment (assafshp) [inline demo/calls/demoApp.js]: if it is only for demo it is ok

File: dist/calls/calls.js [hunk 2]
```diff
@@ -33,11 +33,11 @@
     'use strict';
 
     angular.module('znk.infra.calls')
-        .config(["WebcallSrvProvider", "ENV", function (WebcallSrvProvider, ENV) {
+        .config(["WebcallSrvProvider", function (WebcallSrvProvider) {
             'ngInject';
             WebcallSrvProvider.setCallCred({
-                username: ENV.plivoUsername,
-                password: ENV.plivoPassword
+            username: 'ZinkerzDev160731091034',     // ENV.plivoUsername,
```
Comment (assafshp) [inline dist/calls/calls.js]: should be ENV

File: dist/calls/calls.js [hunk 3]
```diff
@@ -33,11 +33,11 @@
...
+            password: 'zinkerz$9999'     // ENV.plivoPassword
```
Comment (assafshp) [inline dist/calls/calls.js]: env

File: dist/calls/calls.js [hunk 4]
```diff
@@ -105,7 +107,24 @@
                                 var curBtnStatus = modelValue.isOffline ? CallsBtnStatusEnum.OFFLINE_BTN.enum : CallsBtnStatusEnum.CALL_BTN.enum;
                                 receiverId = modelValue.receiverId;
                                 _changeBtnState(curBtnStatus);
-                                _initializeBtnStatus(receiverId);
+                                _initializeBtnStatus(receiverId).then(function() {
```
Comment (assafshp) [inline dist/calls/calls.js]: .CATCH()

Comment (assafshp) [inline dist/calls/calls.js]: _initializeBtnStatus doesn't return any data ?

File: dist/calls/calls.js [hunk 5]
```diff
@@ -105,7 +107,24 @@
...
+                                    if (angular.isDefined(modelValue.toggleAutoCall) && isTeacher) {
+                                        CallsSrv.isUserInActiveCall().then(function (isInActiveCall) {
+                                            switch (modelValue.toggleAutoCall) {
```
Comment (assafshp) [inline dist/calls/calls.js]: .catch

File: dist/calls/calls.js [hunk 6]
```diff
@@ -105,7 +107,24 @@
...
+                                                    if (!isInActiveCall) {
+                                                        vm.clickBtn();
+                                                    }
+                                                    break;
+                                                case autoCallStatusEnum.DISABLE.enum:
+                                                    if (isInActiveCall) {
+                                                        vm.clickBtn();
+                                                    }
+                                                    break;
+                                            }
```
Comment (assafshp) [inline dist/calls/calls.js]: case default not needed

File: dist/calls/calls.js [hunk 7]
```diff
@@ -105,7 +107,24 @@
...
+                                        CallsSrv.isUserInActiveCall().then(function (isInActiveCall) {
+                                            switch (modelValue.toggleAutoCall) {
+                                                case autoCallStatusEnum.ACTIVATE.enum:
+                                                    if (!isInActiveCall) {
+                                                        vm.clickBtn();
+                                                    }
+                                                    break;
+                                                case autoCallStatusEnum.DISABLE.enum:
+                                                    if (isInActiveCall) {
+                                                        vm.clickBtn();
```
Comment (assafshp) [inline dist/calls/calls.js]: looks strange that function will clicked button

File: dist/exerciseResult/exerciseResult.js [hunk 8]
```diff
@@ -380,66 +381,100 @@
             };
 
             /* Module Results Functions */
-            this.getModuleExerciseResult = function (userId, moduleId, exerciseTypeId, exerciseId) {
-
+            this.getModuleExerciseResult = function (userId, moduleId, exerciseTypeId, exerciseId, assignContentType, examId, dontInit) {
+                var dontIntExerciseRes = exerciseTypeId !== ExerciseTypeEnum.SECTION.enum;  // todo - check if it's ok.
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: what is this todo ?

File: dist/exerciseResult/exerciseResult.js [hunk 9]
```diff
@@ -380,66 +381,100 @@
...
                 return $q.all([
-                    this.getExerciseResult(exerciseTypeId, exerciseId, null, null, true),
+                    this.getExerciseResult(exerciseTypeId, exerciseId, examId, null, dontIntExerciseRes),
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: why did you change it to examId ?

File: dist/exerciseResult/exerciseResult.js [hunk 10]
```diff
@@ -380,66 +381,100 @@
...
+                        return USER_MODULE_RESULTS_PATH.replace('$$uid', userId);
+                    case 2:
+                        return USER_HOMEWORK_RESULTS_PATH.replace('$$uid', userId);
+                }
+            }
+
+            this.getModuleResult = function (userId, moduleId, withDefaultResult, withExerciseResults, assignContentType) {
                 return InfraConfigSrv.getStudentStorage().then(function (StudentStorageSrv) {
-                    var userResultsPath = USER_MODULE_RESULTS_PATH.replace('$$uid', userId);
+                    var userResultsPath = _getAssignContentUserPath(userId, assignContentType);
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: good change !

File: dist/exerciseResult/exerciseResult.js [hunk 11]
```diff
@@ -380,66 +381,100 @@
...
-                                moduleResultGuid = defaultResult.guid;
+                            var defaultResult = {};
+                            var moduleResultGuid = moduleResultsGuids[moduleId];
+
+                            if (!moduleResultGuid) {
+                                if (!withDefaultResult) {
+                                    return null;
+                                } else {
+                                    defaultResult = ExerciseResultSrv.getDefaultModuleResult(moduleId, userId);
+                                    moduleResultGuid = defaultResult.guid;
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: what will happen if defaultResult will be undefined ? if it is possible ?

File: dist/exerciseResult/exerciseResult.js [hunk 12]
```diff
@@ -380,66 +381,100 @@
...
+                            return StudentStorageSrv.get(resultPath).then(function (moduleResult) {
+                                var promArray = [];
+                                var exerciseTypeId, exerciseId;
+
+                                if (moduleResult.exercises && withExerciseResults) {
+                                    moduleResult.exerciseResults = [];
+                                    angular.forEach(moduleResult.exercises, function (exerciseData) {
+
+                                        if (angular.isDefined(exerciseData.examId)) {
+                                            exerciseTypeId = ExerciseTypeEnum.SECTION.enum;
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: why do you need this line ? isn't it already defined well in content ?

File: dist/exerciseResult/exerciseResult.js [hunk 13]
```diff
@@ -380,66 +381,100 @@
...
-                                    angular.forEach(exerciseResult, function (exerciseResultGuid, exerciseId) {
-                                        getExerciseResultsProm = getExerciseResultsProm.then(function () {
-                                            return ExerciseResultSrv.getModuleExerciseResult(userId, moduleId, exerciseTypeId, exerciseId).then(function (exerciseResults) {
-                                                if (exerciseResults) {
-                                                    moduleResult.exerciseResults[exerciseTypeId][exerciseId] = exerciseResults;
+                                        var prom = ExerciseResultSrv.getModuleExerciseResult(userId, moduleId, exerciseTypeId, exerciseId, assignContentType, moduleResult.examId).then(function (exerciseResults) {
+                                            if (exerciseResults) {
+                                                if(!moduleResult.exerciseResults[exerciseData.exerciseTypeId]){
+                                                    moduleResult.exerciseResults[exerciseData.exerciseTypeId] = {};
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: why empty object

File: dist/exerciseResult/exerciseResult.js [hunk 14]
```diff
@@ -380,66 +381,100 @@
...
+                                                moduleResult.exerciseResults[exerciseData.exerciseTypeId][exerciseData.exerciseId] = exerciseResults;
+                                            }
                                         });
-                                    });
+                                        promArray.push(prom);
+                                        }
+                                    );
+                                }
+
+                                return $q.all(promArray).then(function () {
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: .catch

File: dist/exerciseResult/exerciseResult.js [hunk 15]
```diff
@@ -525,13 +563,32 @@
                                 dataToSave[exerciseResultsPath].status = exerciseStatuses[exerciseTypeId][exerciseId].status;
                             }
 
+                            var getSectionAggregatedDataProm = $q.when();   // todo - duplicate code. make as a function.
```
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: todo ?

File: dist/presence/presence.js [hunk 16]
```diff
@@ -5,16 +5,17 @@
...
-        .config(["IdleProvider", "KeepaliveProvider", "ENV", function (IdleProvider, KeepaliveProvider, ENV) {
+        .config(["IdleProvider", "KeepaliveProvider", function (IdleProvider, KeepaliveProvider) {
             // userIdleTime: how many sec until user is 'IDLE'
             // idleTimeout: how many sec after idle to stop track the user, 0: keep track
             // idleKeepalive: keepalive interval in sec
 
-            IdleProvider.idle(ENV.userIdleTime || 30);
-            IdleProvider.timeout(ENV.idleTimeout || 0);
-            KeepaliveProvider.interval(ENV.idleKeepalive || 2);
+            IdleProvider.idle(30);
```
Comment (assafshp) [inline dist/presence/presence.js]: why not env?

File: dist/presence/presence.js [hunk 17]
```diff
@@ -5,16 +5,17 @@
...
+            IdleProvider.timeout(0);
+            KeepaliveProvider.interval(2);
         }])
         .run(["PresenceService", "Idle", function (PresenceService, Idle) {
+            'ngInject';
```
Comment (assafshp) [inline dist/presence/presence.js]: good !

File: dist/znkChat/znkChat.js [hunk 18]
```diff
@@ -296,6 +296,7 @@
                         }
 
                         $timeout(function () {
+                            if (!scope.chatterObj.chatMessages) { return; }
```
Comment (assafshp) [inline dist/znkChat/znkChat.js]: why ?

File: src/components/storage/storageSrv.js [hunk 19]
```diff
@@ -233,9 +233,12 @@
 
                 return this.get(path).then(function (pathValue) {
                     self.adapter.onEvent('value', pathValue.$$path, function (serverValue) {
+                        if (typeof serverValue !== 'object'){
+                            $log.error('getAndBindToServer Fn support only object value');
```
Comment (assafshp) [inline src/components/storage/storageSrv.js]: I would have change it to "getAndBindToServer fn supports only binding to an object (and not other value)"

File: Gruntfile.js [hunk 20]
```diff
@@ -435,7 +435,7 @@ module.exports = function (grunt) {
 
     grunt.registerTask('dist', [
         'jshint:all',
-        'karma:build',
+        // 'karma:build',
```
Comment (OfirZinkerz) [inline Gruntfile.js]: not me, probably one of the merge

File: demo/calls/demoApp.js (hunk 1 above)
Comment (OfirZinkerz) [inline demo/calls/demoApp.js]: I test the registerTeacherContext Fn

File: dist/calls/calls.js (hunk 2 above)
Comment (OfirZinkerz) [inline dist/calls/calls.js]: OK

File: dist/calls/calls.js (hunk 3 above)
Comment (OfirZinkerz) [inline dist/calls/calls.js]: OK

File: dist/calls/calls.js (hunk 4 above)
Comment (OfirZinkerz) [inline dist/calls/calls.js]: it return button status, I improve the Fn

File: dist/calls/calls.js (hunk 7 above)
Comment (OfirZinkerz) [inline dist/calls/calls.js]: not my names

File: dist/exerciseResult/exerciseResult.js (hunk 9 above)
Comment (OfirZinkerz) [inline dist/exerciseResult/exerciseResult.js]: oren/kosta

File: dist/exerciseResult/exerciseResult.js (hunk 10 above)
Comment (OfirZinkerz) [inline dist/exerciseResult/exerciseResult.js]: oren/kosta

File: dist/exerciseResult/exerciseResult.js (hunk 11 above)
Comment (OfirZinkerz) [inline dist/exerciseResult/exerciseResult.js]: oren/kosta

File: dist/exerciseResult/exerciseResult.js (hunk 13 above)
Comment (OfirZinkerz) [inline dist/exerciseResult/exerciseResult.js]: to initialise if undefine

File: dist/exerciseResult/exerciseResult.js (hunk 14 above)
Comment (OfirZinkerz) [inline dist/exerciseResult/exerciseResult.js]: added

File: dist/exerciseResult/exerciseResult.js (hunk 15 above)
Comment (OfirZinkerz) [inline dist/exerciseResult/exerciseResult.js]: oren

File: dist/presence/presence.js (hunk 16 above)
Comment (OfirZinkerz) [inline dist/presence/presence.js]: added

File: dist/znkChat/znkChat.js (hunk 18 above)
Comment (OfirZinkerz) [inline dist/znkChat/znkChat.js]: kosta?

File: Gruntfile.js (hunk 20 above)
Comment (assafshp) [inline Gruntfile.js]: please return

File: bower.json [hunk 21]
```diff
@@ -6,6 +6,7 @@
   ],
   "dependencies": {
     "angular": "1.5.8",
+    "all-envs-config": "https://github.com/Knowledge-OTP/all-envs-config.git#zinkerz/dev",
```
Comment (assafshp) [inline bower.json]: should go to ~v0.0.10 and not zinkerz/dev

File: demo/znkExercise/demo.js [hunk 22]
```diff
@@ -12,6 +12,20 @@
             'znk.infra.popUp',
             'demoEnv'
         ])
+        // .config(function (PresenceServiceProvider, SessionSrvProvider, znkAnalyticsSrvProvider, CallsUiSrvProvider) {
+
+        //     PresenceServiceProvider.setAuthServiceName('AuthService');
+
+        //     var calleeNameFunc = function ($q) {
+        //         'ngInject';
```
Comment (assafshp) [inline demo/znkExercise/demo.js]: why ?

File: demoShared/env.service.js [hunk 23]
```diff
@@ -1,7 +1,7 @@
 (function (angular) {
     'use strict';
 
-    angular.module('demoEnv',[]).constant('ENV', {
+    angular.module('demoEnv',[]).value('ENV', {
```
Comment (assafshp) [inline demoShared/env.service.js]: if possible return to constant

File: dist/exerciseResult/exerciseResult.js (hunk 12 above)
Comment (assafshp) [inline dist/exerciseResult/exerciseResult.js]: kosta/oren

File: dist/storage/storage.js [hunk 24]
```diff
@@ -161,6 +161,7 @@
 
                         var ref = this.getRef(path);
                         ref.on(type, function (snapshot) {
+                            if (!self.__registeredEvents[type][path]) { self.__registeredEvents[type][path] = []; }
```
Comment (assafshp) [inline dist/storage/storage.js]: shouldn't execute the entire block

File: Gruntfile.js (hunk 20 above)
Comment (OfirZinkerz) [inline Gruntfile.js]: OK

File: dist/exerciseResult/exerciseResult.js (hunk 9 above)
Comment (kostamas) [inline dist/exerciseResult/exerciseResult.js]: because now the module result can be also exam

File: dist/exerciseResult/exerciseResult.js (hunk 12 above)
Comment (kostamas) [inline dist/exerciseResult/exerciseResult.js]: no - if it's exam, there is no exerciseTypeId
